<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"></ul></div>

The goal is to give a prediction of in which care unit will a patient be in in a certain number of days (or if they will be discharged). This way, hospital workload per care unit/department can be predicted. 

For this, I will have a table of patients with multiple features including medical data (diagnosis, medications...), previous stays and transfers, some demographic data, duration of stay... This can be done using the admissions table and joining information from the diagnosis table (specific per stay) 

At admission, which department will the patient end up in?
Will the patient visit this deparment in their stay?

Make more questions during the weekend.

Do the questions being clear about which variables you will use

And in two weeks have more or less the table ready. 

Before continuing, extract a smaller table that only contains relevant columns for finding atc1s. Find categories and fill them in that table. Then update/rejoin to prescriptions table and find transfer numbers.

I want to predict department occupancy in a hospital and I have a dataset that includes different information from patients. Initially at least, I am simplifying that by predicting instead the department that the patients are going to go to next. 

After a preliminary EDA I saw that I could not use the diagnosis data because it was not dated (it is a table with diagnosis for the whole stay), so now I am working to merge my database with a huge prescriptions table that is dated (so I can relate to patients for different stays and hospital transfers). I am converting the drugs to drug categories and did some frequency encoding, but there was like 20% of missing data, so I was filling that until recently. 

In principle I am working with 14 drug categories, but I have loads of data so could increase that to subcategories and apply PCA. In terms of modeling, my plan is to start with logistic regression and then move on to SVM, random forests...

In [1]:
import numpy as np
import pandas as pd
import copy
import cap_utils
import importlib
importlib.reload(cap_utils);

In [2]:
prescriptions = pd.read_csv('G:\My Drive\Brainstation\Capstone project\MIMIC IV/prescriptions.csv')

C:\Users\Nacho\AppData\Local\Temp\ipykernel_20276\3260518518.py:1: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
  prescriptions = pd.read_csv('G:\My Drive\Brainstation\Capstone project\MIMIC IV/prescriptions.csv')


In [3]:
medrecon = pd.read_csv('MIMIC IV/ed/medrecon.csv')

In [4]:
prescriptions['ndc'] = prescriptions.ndc.astype('Int64')
medrecon['ndc'] = medrecon.ndc.astype('Int64')

In [5]:
display(prescriptions.head())
display(prescriptions.info(show_counts = True))

subject_id   hadm_id  pharmacy_id       poe_id  poe_seq  \
0    10000032  22595853     11700683  10000032-34     34.0   
1    10000032  22595853     14779570  10000032-22     22.0   
2    10000032  22595853     19796602  10000032-50     50.0   
3    10000032  22595853     20256254  10000032-32     32.0   
4    10000032  22595853     28781051  10000032-27     27.0   

             starttime             stoptime drug_type  \
0  2180-05-07 01:00:00  2180-05-07 22:00:00      MAIN   
1  2180-05-07 00:00:00  2180-05-07 22:00:00      MAIN   
2  2180-05-08 08:00:00  2180-05-07 22:00:00      MAIN   
3  2180-05-07 01:00:00  2180-05-07 22:00:00      MAIN   
4  2180-05-07 00:00:00  2180-05-07 22:00:00      MAIN   

                          drug formulary_drug_cd     gsn          ndc  \
0                Acetaminophen           APAP500  004490    904198861   
1  Sodium Chloride 0.9%  Flush         NACLFLUSH     NaN            0   
2                   Furosemide            FURO40  008209  51079007320   
3                  Raltegravir           RALT400  063231      6022761   
4                      Heparin            HEPA5I  006549  63323026201   

               prod_strength form_rx dose_val_rx dose_unit_rx form_val_disp  \
0               500mg Tablet     NaN         500           mg             1   
1              10 mL Syringe     NaN           3           mL           0.3   
2                40mg Tablet     NaN          40           mg             1   
3              400 mg Tablet     NaN         400           mg             1   
4  5000 Units / mL- 1mL Vial     NaN        5000         UNIT             1   

  form_unit_disp  doses_per_24_hrs  route  
0            TAB               NaN  PO/NG  
1            SYR               3.0     IV  
2            TAB               1.0  PO/NG  
3            TAB               2.0     PO  
4             mL               3.0     SC

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16219412 entries, 0 to 16219411
Data columns (total 20 columns):
 #   Column             Non-Null Count     Dtype  
---  ------             --------------     -----  
 0   subject_id         16219412 non-null  int64  
 1   hadm_id            16219412 non-null  int64  
 2   pharmacy_id        16219412 non-null  int64  
 3   poe_id             16091738 non-null  object 
 4   poe_seq            16091738 non-null  float64
 5   starttime          16201857 non-null  object 
 6   stoptime           16195756 non-null  object 
 7   drug_type          16219412 non-null  object 
 8   drug               16219411 non-null  object 
 9   formulary_drug_cd  16199404 non-null  object 
 10  gsn                14342210 non-null  object 
 11  ndc                16193185 non-null  Int64  
 12  prod_strength      16212752 non-null  object 
 13  form_rx            20731 non-null     object 
 14  dose_val_rx        16212798 non-null  object 
 15  dose_unit_rx 

None

In [6]:
prescriptions['starttime'] = pd.to_datetime(prescriptions.starttime, infer_datetime_format=True)

In [7]:
display(medrecon.head())
display(medrecon.info(show_counts = True))

subject_id   stay_id            charttime  \
0    10000032  32952584  2180-07-22 17:26:00   
1    10000032  32952584  2180-07-22 17:26:00   
2    10000032  32952584  2180-07-22 17:26:00   
3    10000032  32952584  2180-07-22 17:26:00   
4    10000032  32952584  2180-07-22 17:26:00   

                                name    gsn          ndc  etc_rn  etccode  \
0                  albuterol sulfate  28090  21695042308       1   5970.0   
1                  calcium carbonate   1340  10135021101       1    733.0   
2       cholecalciferol (vitamin D3)  65241  37205024678       1    670.0   
3  emtricitabine-tenofovir [Truvada]  57883  35356007003       1   5849.0   
4          fluticasone [Flovent HFA]  21251  49999061401       1    371.0   

                                      etcdescription  
0  Asthma/COPD Therapy - Beta 2-Adrenergic Agents...  
1    Minerals and Electrolytes - Calcium Replacement  
2                           Vitamins - D Derivatives  
3  Antiretroviral - Nucleoside and Nucleotide Ana...  
4  Asthma Therapy - Inhaled Corticosteroids (Gluc...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3143791 entries, 0 to 3143790
Data columns (total 9 columns):
 #   Column          Non-Null Count    Dtype  
---  ------          --------------    -----  
 0   subject_id      3143791 non-null  int64  
 1   stay_id         3143791 non-null  int64  
 2   charttime       3143791 non-null  object 
 3   name            3143791 non-null  object 
 4   gsn             3143791 non-null  int64  
 5   ndc             3143791 non-null  Int64  
 6   etc_rn          3143791 non-null  int64  
 7   etccode         3131412 non-null  float64
 8   etcdescription  3131412 non-null  object 
dtypes: Int64(1), float64(1), int64(4), object(3)
memory usage: 218.9+ MB


None

Warning about ndc having mixed types. Instead of using isna to find about null values, we will just take as valid NDCs those that are larger that have at least 6 digits (which is the smallest NDC according to the FDA liink). We can also check whether we have other NDC numbers with fewer digits, as these could be fixable mistakes.

Prescriptions and medrecon are similar,thankfully, although instead of hadm_id we have stay_id. And no null values on our drugs column, yay! We lack the drug type column.

In [8]:
def digit_checker(df, d_min = 0, d_max = 6):
    for i in range(d_min,d_max):
        d_no = 10**i

        validndc_bool = df.ndc>=d_no
        print(f'% of NDCs >= 10^{i}:', round(sum(validndc_bool)/len(df)*100,2))
        print('Number of NDCs:', df.ndc[validndc_bool].nunique(),'\n')

In [9]:
display(prescriptions.nunique())
digit_checker(prescriptions,d_max = 12)

subject_id             166758
hadm_id                384772
pharmacy_id          14237899
poe_id               12844633
poe_seq                 16298
starttime              819017
stoptime               795665
drug_type                   3
drug                     9826
formulary_drug_cd        3819
gsn                      7517
ndc                      5745
prod_strength            7326
form_rx                    87
dose_val_rx              4763
dose_unit_rx              146
form_val_disp            5263
form_unit_disp            114
doses_per_24_hrs           17
route                      99
dtype: int64

TypeError: type NAType doesn't define __round__ method

In [10]:
display(medrecon.nunique())
digit_checker(medrecon,d_max = 12)

subject_id        148919
stay_id           323582
charttime         581453
name               19165
gsn                 9262
ndc                 9398
etc_rn                 5
etccode             1208
etcdescription      1212
dtype: int64

% of NDCs >= 10^0: 99.4
Number of NDCs: 9397 

% of NDCs >= 10^1: 99.4
Number of NDCs: 9397 

% of NDCs >= 10^2: 99.4
Number of NDCs: 9397 

% of NDCs >= 10^3: 99.4
Number of NDCs: 9397 

% of NDCs >= 10^4: 99.4
Number of NDCs: 9397 

% of NDCs >= 10^5: 99.4
Number of NDCs: 9397 

% of NDCs >= 10^6: 99.4
Number of NDCs: 9397 

% of NDCs >= 10^7: 99.26
Number of NDCs: 9232 

% of NDCs >= 10^8: 98.91
Number of NDCs: 8902 

% of NDCs >= 10^9: 98.56
Number of NDCs: 8531 

% of NDCs >= 10^10: 98.52
Number of NDCs: 8450 

% of NDCs >= 10^11: 0.0
Number of NDCs: 0 



So all NDCs larger than 1 appear to be correct. There are 12.44% missing NDC, but we will not be filling these. We are interested on the ATR4 classification, so we will match the available NDCs to this class and then try to fill any empty values for that feature.

For the NDC matching we are using a function provided by tal that queries tal tal. The prescriptions table was used as input for this function and the output is uploaded and mapped into our table here.

In [11]:
prescriptions.drop(columns = ['prod_strength','form_rx','dose_val_rx','dose_unit_rx','form_val_disp',\
                              'form_unit_disp','doses_per_24_hrs','pharmacy_id','poe_id','poe_seq','stoptime'],inplace = True)


Dropping some of the columns as they (may) have no relevance as we are substituting the drug features by their categories. This will also allow us to combine our dfs.

In [12]:
print(sum(medrecon.duplicated()))
print(sum(medrecon.drop(columns = ['etccode','etc_rn','etcdescription']).duplicated()))

0
410218


Medications are duplicated to include the different etc codes for one drug. That is okay, we could create dummies and group the columns. However, we will not use the edc in principle. We are going to move to ATCs, which are more interpretable, general and well-recognised. More importantly, they have a very customisable granularity, as we can select the level. At level 0 we have only 14 features, which simplifies our model greatly. We will start with that and increase granularity if necessary.

In [13]:
medrecon.drop(columns =  ['etccode','etc_rn','etcdescription'], inplace = True)
medrecon.drop_duplicates(inplace = True)
medrecon.rename(columns = {'charttime':'starttime','name':'drug'},inplace = True)

In [14]:
prescriptions

subject_id   hadm_id           starttime drug_type  \
0           10000032  22595853 2180-05-07 01:00:00      MAIN   
1           10000032  22595853 2180-05-07 00:00:00      MAIN   
2           10000032  22595853 2180-05-08 08:00:00      MAIN   
3           10000032  22595853 2180-05-07 01:00:00      MAIN   
4           10000032  22595853 2180-05-07 00:00:00      MAIN   
...              ...       ...                 ...       ...   
16219407    19999987  23865745 2145-11-03 00:00:00      MAIN   
16219408    19999987  23865745 2145-11-04 10:00:00      MAIN   
16219409    19999987  23865745 2145-11-10 10:00:00      MAIN   
16219410    19999987  23865745 2145-11-09 10:00:00      MAIN   
16219411    19999987  23865745 2145-11-09 10:00:00      MAIN   

                                 drug formulary_drug_cd     gsn          ndc  \
0                       Acetaminophen           APAP500  004490    904198861   
1         Sodium Chloride 0.9%  Flush         NACLFLUSH     NaN            0   
2                          Furosemide            FURO40  008209  51079007320   
3                         Raltegravir           RALT400  063231      6022761   
4                             Heparin            HEPA5I  006549  63323026201   
...                               ...               ...     ...          ...   
16219407      Influenza Virus Vaccine         INFL0.5LF  069637  33332001201   
16219408         Levothyroxine Sodium            LEVO50  006649     74455211   
16219409                   Duloxetine            DULO30  057892      2324033   
16219410               Venlafaxine XR         VENL150XR  046405  68084048601   
16219411                   Duloxetine            DULO30  057892      2324033   

          route  
0         PO/NG  
1            IV  
2         PO/NG  
3            PO  
4            SC  
...         ...  
16219407     IM  
16219408  PO/NG  
16219409     PO  
16219410     PO  
16219411     PO  

[16219412 rows x 9 columns]

In [15]:
cap_utils.display_sbs(['Stay_id from medrecon','Hadm_id from prescriptions'],\
                      medrecon.stay_id.describe(),prescriptions.hadm_id.describe())

Stay_id from medrecon 
 
 
   
 stay_id 
 
 
 
 
 count 
 2733573.000000 
 
 
 mean 
 35001782.253237 
 
 
 std 
 2885614.265219 
 
 
 min 
 30000012.000000 
 
 
 25% 
 32502391.000000 
 
 
 50% 
 35000450.000000 
 
 
 75% 
 37498150.000000 
 
 
 max 
 39999964.000000 
 
 
 

 
 Hadm_id from prescriptions 
 
 
   
 hadm_id 
 
 
 
 
 count 
 16219412.000000 
 
 
 mean 
 25004407.280042 
 
 
 std 
 2884617.408397 
 
 
 min 
 20000019.000000 
 
 
 25% 
 22514037.000000 
 
 
 50% 
 25008728.000000 
 
 
 75% 
 27498545.000000 
 
 
 max 
 29999928.000000

Lovely, comparable, no overlap. We will combine them, adding a new column to indicate rows where hadm_id is actually stay_id (for info not for modelling).

In [16]:
prescriptions['stay_id_ind'] = 0
medrecon['stay_id_ind'] = 1
prescriptions = pd.concat([prescriptions,medrecon.rename(columns = {'stay_id':'hadm_id'})],axis = 0)
prescriptions.reset_index(inplace = True,drop = True)

In [17]:
cap_utils.display_sbs(['NDCs from hospital','NDCs from ED'],\
                      prescriptions.ndc[prescriptions.stay_id_ind == 0].describe(),\
                      prescriptions.ndc[prescriptions.stay_id_ind == 1].describe())

NDCs from hospital 
 
 
   
 ndc 
 
 
 
 
 count 
 16193185.000000 
 
 
 mean 
 17356329121.561031 
 
 
 std 
 26353062677.424332 
 
 
 min 
 0.000000 
 
 
 25% 
 172375910.000000 
 
 
 50% 
 409707714.000000 
 
 
 75% 
 50383077933.000000 
 
 
 max 
 99207024005.000000 
 
 
 

 
 NDCs from ED 
 
 
   
 ndc 
 
 
 
 
 count 
 2733573.000000 
 
 
 mean 
 16184444403.336649 
 
 
 std 
 11891976106.486269 
 
 
 min 
 0.000000 
 
 
 25% 
 10337016215.000000 
 
 
 50% 
 10939053344.000000 
 
 
 75% 
 16590031330.000000 
 
 
 max 
 99207027675.000000

It seems like ED and the hospital have completely different suppliers, because there is little overlap of ndcs. Let's export the unique ndc list to query the ATCs.

In [18]:
np.char.zfill(prescriptions.ndc.unique().astype('str'),11)

array(['00904198861', '00000000000', '51079007320', ..., '10939016244',
       '11917010387', '43376031060'], dtype='<U11')

In [19]:
pd.Series(name = 'NDC',data= np.char.zfill(prescriptions.ndc.unique().astype('str'),11)).to_csv('unique_ndcs.csv')

In [20]:
mapped_NDC = pd.read_csv(r'G:\My Drive\Brainstation\Capstone project\ndc_map\output\atc4\ndc_map 2022_10_12 17_12 (atc4).csv',usecols = ['ndc','atc4','atc4_name'])

In [21]:
mapped_NDC.dropna(inplace = True)
mapped_NDC.ndc = mapped_NDC.ndc.astype('int64')

In [22]:
mapped_NDC.drop_duplicates(inplace = True)
len(mapped_NDC)

21710

In [23]:
mapped_NDC.head()

ndc   atc4                                  atc4_name
2  2050301  J01GB                      Other aminoglycosides
3  2050301  S01AA                                Antibiotics
4  2060440  J04AB                                Antibiotics
5  2100902  R05DA            Opium alkaloids and derivatives
8  2143301  A10BJ  Glucagon-like peptide-1 (GLP-1) analogues

In [24]:
mapped_NDC.atc4_name.nunique()

518

A lot of categories. Feasible, but we will first try taking just the first letter (lvl 1 of classification) of the atc_4 instead

In [25]:
atc1 = mapped_NDC[['ndc','atc4']].copy()
atc1['atc4'] = atc1.atc4.str[0]
atc1.rename(columns = {'atc4':'atc1'},inplace = True)
atc1.drop_duplicates(inplace = True)

In [26]:
atc1.head()
print(len(atc1))
print(atc1.nunique())
display(atc1.groupby('ndc').nunique().describe())

16937
ndc     10770
atc1       14
dtype: int64


atc1
count  10770.000000
mean       1.572609
std        1.181846
min        1.000000
25%        1.000000
50%        1.000000
75%        2.000000
max        8.000000

In [27]:
atc1 = pd.concat([atc1,pd.get_dummies(atc1.atc1,prefix = None)],axis = 1).drop(columns = 'atc1')
atc1 = atc1.groupby('ndc').any()*1
display(atc1.head())
print(atc1.shape)

A  B  C  D  G  H  J  L  M  N  P  R  S  V
ndc                                              
2050301  0  0  0  0  0  0  1  0  0  0  0  0  1  0
2060440  0  0  0  0  0  0  1  0  0  0  0  0  0  0
2100902  0  0  0  0  0  0  0  0  0  0  0  1  0  0
2143301  1  0  0  0  0  0  0  0  0  0  0  0  0  0
2143401  1  0  0  0  0  0  0  0  0  0  0  0  0  0

(10770, 14)


There is more than one atc1 per ndc code! We were going to use dummy variables in the future, so we will implement those now before the merge. That way we will have just one category per ndc. Also note that not all ndcs have been mapped, 490 out of 5744 are still unclassified. We will try to fill those after the merge.

In [28]:
prescriptions.reset_index(inplace = True)
prescriptions.rename(columns = {'index':'join_on_index'},inplace = True)
prescriptions.set_index('ndc',inplace = True)
cat_pres = prescriptions[['join_on_index','drug_type','drug','formulary_drug_cd','gsn']].join(atc1)
cat_pres.reset_index(inplace = True)

In [29]:
display(cat_pres.sample(10))
print('Missing categories %:',sum(cat_pres.A.isna())/len(cat_pres)*100)
cat = list('ABCDGHJLMNPRSV')
cat_pres[cat] = cat_pres[cat].fillna(-1).astype('int8')

print(cat_pres.info(show_counts = True))

ndc  join_on_index drug_type                         drug  \
2736426      29152611       13509589      MAIN  Mupirocin Nasal Ointment 2%   
17333385  63323026201       12321938      MAIN                      Heparin   
17628465  63323054011       14636529      MAIN       Heparin (Hemodialysis)   
7175213     409166010        5617397      MAIN              Dexmedetomidine   
15356649  51079045120       11050356      MAIN                   amLODIPine   
1959132             0       15957349      BASE                           NS   
7436797     409427701       12779349      MAIN                 Lidocaine 2%   
18201848  66689036430       10713077      MAIN             Milk of Magnesia   
9698526     904198861         758989      MAIN                Acetaminophen   
1608893             0       13089411      BASE                          D5W   

         formulary_drug_cd     gsn    A    B    C    D    G    H    J    L  \
2736426           BACT2/1O  018370  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   
17333385            HEPA5I  006549  0.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0   
17628465            HEP10I  006543  0.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0   
7175213           DEXM400I  070877  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
15356649             AMLO5  016926  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0   
1959132           METRBASE     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
7436797           LIDO220I  003405  0.0  0.0  1.0  1.0  0.0  0.0  0.0  0.0   
18201848            MOM30L  003026  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
9698526            APAP500  004490  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1608893            HEPBASE     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

            M    N    P    R    S    V  
2736426   0.0  0.0  0.0  1.0  0.0  0.0  
17333385  0.0  0.0  0.0  0.0  1.0  0.0  
17628465  0.0  0.0  0.0  0.0  1.0  0.0  
7175213   0.0  1.0  0.0  0.0  0.0  0.0  
15356649  0.0  0.0  0.0  0.0  0.0  0.0  
1959132   NaN  NaN  NaN  NaN  NaN  NaN  
7436797   0.0  1.0  0.0  1.0  1.0  0.0  
18201848  NaN  NaN  NaN  NaN  NaN  NaN  
9698526   0.0  1.0  0.0  0.0  0.0  0.0  
1608893   NaN  NaN  NaN  NaN  NaN  NaN

Missing categories %: 18.4277516180169
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18952985 entries, 0 to 18952984
Data columns (total 20 columns):
 #   Column             Non-Null Count     Dtype 
---  ------             --------------     ----- 
 0   ndc                18926758 non-null  Int64 
 1   join_on_index      18952985 non-null  int64 
 2   drug_type          16219412 non-null  object
 3   drug               18952984 non-null  object
 4   formulary_drug_cd  16199404 non-null  object
 5   gsn                17075783 non-null  object
 6   A                  18952985 non-null  int8  
 7   B                  18952985 non-null  int8  
 8   C                  18952985 non-null  int8  
 9   D                  18952985 non-null  int8  
 10  G                  18952985 non-null  int8  
 11  H                  18952985 non-null  int8  
 12  J                  18952985 non-null  int8  
 13  L                  18952985 non-null  int8  
 14  M                  18952985 non-null  int

In [30]:
cat_pres['ndc'] = cat_pres.ndc.floordiv(100)
cat_pres.loc[cat_pres.ndc == 0,'ndc'] = pd.NA
cat_pres['ndc'] = cat_pres.ndc.astype('Int64')

Note that the new category dummy columns are floats after the join, not int8s. We have filled the NaNs as -1 and changed the data type of the columns. This is useful to decrease the size of the dataframe and it will also be useful to fill the missing values (which is 19.26% of the rows), as we will see below. We will fill these missing values by matching with other columns and we there are a few candidates column for this: drug, pharmacy_id, gsn and formulary_drug_cd. The goal will be to find the column with values shared by the largest amount of rows, so matching ATR1s is more likely. 

In [31]:
nndrug = cat_pres[cat_pres.drug.isna()].index.values[0]
display(cat_pres.loc[nndrug])

display(cat_pres[cat_pres.ndc == cat_pres.ndc.loc[nndrug]].head(1))

cat_pres.loc[nndrug,['drug', 'formulary_drug_cd']] = ('Simvastatin','SIMV40')

print('Value counts of drugs:\n',cat_pres.drug.value_counts())
cat_pres.drug = cat_pres.drug.str.lower()
drugs_vc = cat_pres.drug.value_counts()

clin_drug = cat_pres.drug.str.contains('clindamyci')
display(cat_pres[clin_drug].head(1))
cat_pres.drug[clin_drug] = 'clindamycin'

print('\n\nValue counts of drugs after making all letters lowercase:\n',drugs_vc)
print('\nDescription:\n',drugs_vc.describe())

print('\nNumber of drugs used only once:', sum(drugs_vc>1))

cat_pres['drug'] = cat_pres.drug.str.replace('[^a-zA-Z]','', regex = True)
drugs_vc = cat_pres.drug.value_counts()

print('\n\nValue counts of drugs after making all letters lowercase:\n',drugs_vc)
print('\nDescription:\n',drugs_vc.describe())

print('\nNumber of drugs used only once:', sum(drugs_vc>1))



ndc                  510790456
join_on_index         16133454
drug_type                 MAIN
drug                       NaN
formulary_drug_cd      SIMVIND
gsn                     016579
A                            0
B                            0
C                            1
D                            0
G                            0
H                            0
J                            0
L                            0
M                            0
N                            0
P                            0
R                            0
S                            0
V                            0
Name: 15420049, dtype: object

ndc  join_on_index drug_type         drug formulary_drug_cd  \
15400176  510790456            351      MAIN  Simvastatin            SIMV40   

             gsn  A  B  C  D  G  H  J  L  M  N  P  R  S  V  
15400176  016579  0  0  1  0  0  0  0  0  0  0  0  0  0  0

Value counts of drugs:
 Insulin                         631400
0.9% Sodium Chloride            606705
Sodium Chloride 0.9%  Flush     544798
Potassium Chloride              501615
Acetaminophen                   465254
                                 ...  
niferex                              1
iron complex                         1
*tasabri infusion                    1
hyoscyamine sulfate [Hyomax]         1
Lumigan                              1
Name: drug, Length: 25853, dtype: int64


ndc  join_on_index drug_type             drug formulary_drug_cd gsn  \
1992507  <NA>       16389120       NaN  *clindamycin 1%               NaN   0   

         A  B  C  D  G  H  J  L  M  N  P  R  S  V  
1992507 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1

C:\Users\Nacho\AppData\Local\Temp\ipykernel_20276\2913060166.py:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cat_pres.drug[clin_drug] = 'clindamycin'




Value counts of drugs after making all letters lowercase:
 insulin                                                    631424
0.9% sodium chloride                                       606705
sodium chloride 0.9%  flush                                544798
potassium chloride                                         508257
acetaminophen                                              496274
                                                            ...  
nebulizer & compressor [comp-air elite comp neb system]         1
*orajel                                                         1
*hiv meds x2                                                    1
*beqyree                                                        1
lumigan                                                         1
Name: drug, Length: 22976, dtype: int64

Description:
 count     22976.000000
mean        824.903595
std       11995.582672
min           1.000000
25%           1.000000
50%           3.000000
75%          17.0000

In [32]:
empty_drug = cat_pres.drug ==u''
print(f'There are {sum(empty_drug)} empty drugs')
display(cat_pres[empty_drug].sample())

fdrugs_empty = cat_pres.formulary_drug_cd[empty_drug].drop_duplicates().values
empty_drug = cat_pres.formulary_drug_cd.isin(fdrugs_empty)

cat_pres.drug.where(~empty_drug,cat_pres.formulary_drug_cd,inplace = True)

print(f'{sum(drugs_vc == 1)/len(cat_pres)*100}% of values in drugs column appears only once')

There are 113 empty drugs


ndc  join_on_index drug_type drug formulary_drug_cd gsn  A  B  C  D  \
2005356  <NA>       18349221       NaN                    NaN   0 -1 -1 -1 -1   

         G  H  J  L  M  N  P  R  S  V  
2005356 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1

0.039444973971118534% of values in drugs column appears only once


There was just one missing value for the drug column, which was easily filled by checking elsewhere in the table. After this, we analysed the number of rows for each drug. Interestingly, the formulary_drug_code was different than the original for that row. 

The input of these fields does not seem consistent. For example "Clindamycin" is misspelt (fixed for this particular case), abbreviations may not be consistent and upper and lower case seem to be used with little care. This last issue could be fixed with str.lower(), though. 

This means too many groups so matching categories will be harder. Let us look at the other columns with drug identifiers.

In [33]:
print('gsn value counts:\n',cat_pres.gsn.value_counts())

fd_vc = cat_pres.formulary_drug_cd.value_counts()
print('\n\nformulary_drug_cd value counts:\n',fd_vc)
print('\nDescription:\n',fd_vc.describe())

fd_vc_5 = fd_vc[fd_vc<=5].value_counts()
print('\nNumber of formulary_drug_cd present on only 5 rows or less:\n',fd_vc_5)
print('\n% of rows with drugs used 5 times or less:', round(fd_vc_5.sum()/fd_vc.sum()*100,5))

gsn value counts:
 001210                          816986
006549                          330441
027413                          286850
004489                          264211
003009                          259342
                                 ...  
74684                                1
79087                                1
63452                                1
63444                                1
021723 021726 021725 021724          1
Name: gsn, Length: 15625, dtype: int64


formulary_drug_cd value counts:
 NACLFLUSH    544603
HEPA5I       330717
ACET325      265279
MAG2PM       265117
DOCU100      260339
              ...  
TTOX5I            1
POLIOV            1
RANI50PM          1
LEUC200I          1
HALC0.1C          1
Name: formulary_drug_cd, Length: 3819, dtype: int64

Description:
 count      3819.000000
mean       4241.792092
std       19845.938890
min           1.000000
25%          10.000000
50%          99.000000
75%        1150.000000
max      544603.000000
Name: f

From these candidates, formulary_drug_cd, gsn, and drugs seems the best to find the right ATR1s, but they have many null values. Additionally, the gsn column has 1.8 M missing values and sometimes several codes in one field. We will have to work around this
The pharmacy_id has many unique values, so this variable is not specific to the same drug, just an inventory code specific per item.  

We want to decrease the number of unique values as this will increase our chances of filling missing ATR1 fields. All our columns have many null values, but we can recursively fill categories based on all the identifying columns.

In [34]:
len(cat_pres)

18952985

In [35]:
cat_pres[cat].sum(axis = 1).value_counts()

 1     9948066
-14    3492609
 2     3036333
 3     1624373
 4      440099
 5      301478
 6       88324
 7       21116
 8         587
dtype: int64

In [36]:
cat_pres.gsn.value_counts()

001210                          816986
006549                          330441
027413                          286850
004489                          264211
003009                          259342
                                 ...  
74684                                1
79087                                1
63452                                1
63444                                1
021723 021726 021725 021724          1
Name: gsn, Length: 15625, dtype: int64

In [37]:
cap_utils.rec_cat_filler(cat_pres,['gsn','drug','formulary_drug_cd','ndc'])

% of missing values before loop: 18.4277516180169
gsn column split, starting loop...
Dataframe exploded, starting for loop #1

Loop #1, missing values % after filling with gsn column: 17.431904395286992
Loop #1, missing values % after filling with drug column: 10.667095422276127
Loop #1, missing values % after filling with formulary_drug_cd column: 9.983328100178527
Loop #1, missing values % after filling with ndc column: 9.78836216883194

Loop #1, missing values % after re-imploding df: 9.760520572353116
Dataframe exploded, starting for loop #2

Loop #2, missing values % after filling with gsn column: 9.737872354487646
Loop #2, missing values % after filling with drug column: 8.976494373529803
Loop #2, missing values % after filling with formulary_drug_cd column: 8.962797136767094
Loop #2, missing values % after filling with ndc column: 8.955551230112366

Loop #2, missing values % after re-imploding df: 8.958145643021401
Dataframe exploded, starting for loop #3

Loop #3, missing value

In [38]:
cap_utils.display_sbs(['Drugs value counts']*2,\
            cat_pres.drug[cat_pres.A<0].value_counts(normalize = True).iloc[0:25]*100,\
            cat_pres.drug[cat_pres.A<0].value_counts(normalize = True).iloc[25:50]*100)

Drugs value counts 
 
 
   
 drug 
 
 
 
 
 bag 
 19.861040 
 
 
 isoosmoticdextrose 
 19.520665 
 
 
 vial 
 6.872297 
 
 
 polyethyleneglycol 
 6.466611 
 
 
 influenzavaccinequadrivalent 
 5.359550 
 
 
 neutraphos 
 4.704206 
 
 
 influenzavirusvaccine 
 4.623353 
 
 
 simethicone 
 3.111876 
 
 
 soln 
 2.821791 
 
 
 pneumococcalvacpolyvalent 
 2.641478 
 
 
 potassiumphosphate 
 2.239080 
 
 
 pneumococcalvalentpolysaccharidevaccine 
 2.062653 
 
 
 syringe 
 1.683943 
 
 
 phosphorus 
 1.550707 
 
 
 epoetinalfa 
 1.491226 
 
 
 sarnalotion 
 1.216609 
 
 
 syringesodiumchloride 
 1.216310 
 
 
 readicatbariumsulfatesuspension 
 1.107061 
 
 
 tiotropiumbromide 
 0.972704 
 
 
 prismasatebk 
 0.933548 
 
 
 sodiumpolystyrenesulfonate 
 0.768105 
 
 
 yellowcaddcassette 
 0.597133 
 
 
 vasopressin 
 0.576135 
 
 
 ferricgluconate 
 0.435277 
 
 
 syringeisoosmoticdextrose 
 0.427879 
 
 
 

 
 Drugs value counts 
 
 
   
 drug 
 
 
 
 
 leucovorincalcium 
 0.384015 
 
 
 tuberculinprotein 
 0.347848 
 
 
 gastroviewdiatrizoatemegluminesodium 
 0.301219 
 
 
 lithiumcarbonate 
 0.248388 
 
 
 syringechemo 
 0.241588 
 
 
 gammagardliquid 
 0.231949 
 
 
 citratedextroseacdacrrt 
 0.224850 
 
 
 psylliumwafer 
 0.218124 
 
 
 epiduralbag 
 0.176801 
 
 
 caphosol 
 0.172093 
 
 
 vitaminbcomplex 
 0.170599 
 
 
 hydrocerin 
 0.169328 
 
 
 fosaprepitant 
 0.167909 
 
 
 tbsyringe 
 0.161931 
 
 
 psyllium 
 0.153711 
 
 
 levalbuterolneb 
 0.138766 
 
 
 aquaphorointment 
 0.133684 
 
 
 amp 
 0.124717 
 
 
 psylliumpowder 
 0.118739 
 
 
 mcgvial 
 0.108651 
 
 
 bottle 
 0.107082 
 
 
 syringesterilewater 
 0.104616 
 
 
 bluecaddcassette 
 0.090194 
 
 
 protaminesulfate 
 0.089895 
 
 
 fatemulsion 
 0.080778

Lots of vaccines, let us fill those first

In [39]:
cat = list('ABCDGHJLMNPRSV')
vaccines_bool = (cat_pres.drug.str.contains('vaccine|polyvalent|measlesmumpsrubellavac',regex = True, na = False))
vaccines_bool *= cat_pres.A<0
cat_pres.loc[vaccines_bool,cat] = 0
cat_pres.loc[vaccines_bool,'J'] = 1

print('% of missing values:',cap_utils.cat_check(cat_pres)*100)

% of missing values: 7.907334913207603


Let us have a look at the missing values and the drugs per drug type (from the prescriptions table)

In [40]:
print('% of missing values for MAIN drugs',cap_utils.cat_check(cat_pres[cat_pres.drug_type =='MAIN'])*100)
print('% of missing values for BASE drugs',cap_utils.cat_check(cat_pres[cat_pres.drug_type =='BASE'])*100)
print('% of missing values for ADDITIVE drugs',cap_utils.cat_check(cat_pres[cat_pres.drug_type =='ADDITIVE'])*100)

base_drugs_bool = cat_pres.drug_type == 'BASE'
print('\n',cat_pres.drug[base_drugs_bool].drop_duplicates().values)

% of missing values for MAIN drugs 3.061362940335017
% of missing values for BASE drugs 26.45541596208607
% of missing values for ADDITIVE drugs 4.861388511624192

 ['sodiumchloride' 'isoosmoticdextrose' 'sw' 'dextrose' 'bag' 'vial'
 'isoosmoticsodiumchloride' 'ns' 'isotonicsodiumchloride'
 'syringesodiumchloride' 'soln' 'dw' 'syringe' 'sterilewater'
 'syringechemo' 'mcgvial' 'yellowcaddcassette' 'syringeisoosmoticdextrose'
 'epiduralbag' 'amp' 'bluecaddcassette' 'nssyringeintrapleural'
 'syringesterilewater' 'syringelr' 'syringesw' 'nsepiduralbag'
 'clickchangetochooseabasesolution' 'sodiumchlorideexcelbag' 'nssyringe'
 'prismasatebk' 'sodiumchlyellowcadd' 'syringesns' 'bottle' 'syringens'
 'solution' 'sodiumchloridepfsyringe' 'aerogensyringens' 'syringepca'
 'syringeintraperitoneal' 'syringepd' 'dwglassbottle' 'aerogensyringesw'
 'dwwalbumin' 'tbsyringe' 'viaflexbag' 'sodiumcitrate'
 'syringeintraventricular' 'caddcassette' 'syringeinhalation'
 'syringeintrathecal' 'nsintrathecalbag'

Rows with drug_type "BASE" correspond with procedures (bag, vial...), solvents (sodium chloride, dextrose) and treatments/drugs (lipid/fat emulsions, travasol, amino acids...) apparently using IV therapy. Many of these procedures will not be very specific per hospital department (sodium chloride, bags, dextrose...), but frequency of procedures could be interesting so registering these events in our table could be useful if we use frequency encoding instead of one-hot encoding. Given that most of these events involve IV therapy and solvents are classed typically both as B (Blood products) and as V (Various) in the ATC system, these prescription events have been classified as such. 

In [41]:
base_drugs_bool = base_drugs_bool * cat_pres.A<0 
cat_pres.loc[base_drugs_bool,cat] = 0

cat_pres.loc[base_drugs_bool,'V'] = 1
cat_pres.loc[base_drugs_bool,'B'] = 1
cat_pres.A.value_counts(normalize = True)*100

print('% of missing values:',cap_utils.cat_check(cat_pres)*100)

% of missing values: 4.02603072814124


We still have 4.03% of missing values in our atc1 columns. Let us run the recursive filler function again.


In [42]:
cap_utils.rec_cat_filler(cat_pres,['gsn','drug','formulary_drug_cd','ndc'])

% of missing values before loop: 4.02603072814124
gsn column split, starting loop...
Dataframe exploded, starting for loop #1

Loop #1, missing values % after filling with gsn column: 4.020052165265818
Loop #1, missing values % after filling with drug column: 3.9610482790498107
Loop #1, missing values % after filling with formulary_drug_cd column: 3.9610482790498107
Loop #1, missing values % after filling with ndc column: 3.9609009403524014

Loop #1, missing values % after re-imploding df: 3.952475032297023
Dataframe exploded, starting for loop #2

Loop #2, missing values % after filling with gsn column: 3.9609009403524014
Loop #2, missing values % after filling with drug column: 3.9609009403524014
Loop #2, missing values % after filling with formulary_drug_cd column: 3.9609009403524014
Loop #2, missing values % after filling with ndc column: 3.9609009403524014

Loop #2, missing values % after re-imploding df: 3.952475032297023


We can try and fill manually the drugs that occur most frequently.We can do another round of recursive filling

https://go.drugbank.com/drugs/DB09512

In [43]:
missing_drugs = cat_pres.drug[cat_pres.A<0].value_counts(normalize = True).head(30)
display(missing_drugs*100)
print(f'We will fill {round(missing_drugs.sum()*100,2)}% of the remaining missing drugs manually')
missing_drugsl = missing_drugs.index.values

polyethyleneglycol                      22.206882
neutraphos                              16.154636
simethicone                             10.686443
potassiumphosphate                       7.689189
phosphorus                               5.325259
epoetinalfa                              5.120994
sarnalotion                              4.177936
readicatbariumsulfatesuspension          3.801740
tiotropiumbromide                        3.340347
sodiumpolystyrenesulfonate               2.637738
vasopressin                              1.978496
ferricgluconate                          1.494778
leucovorincalcium                        1.318741
tuberculinprotein                        1.194539
gastroviewdiatrizoatemegluminesodium     1.034412
lithiumcarbonate                         0.852986
gammagardliquid                          0.796531
psylliumwafer                            0.749057
caphosol                                 0.590983
vitaminbcomplex                          0.585850


We will fill 95.8% of the remaining missing drugs manually


In [44]:
cat_pres.loc[cat_pres.drug.isin(missing_drugsl),cat] = 0

cat_pres.loc[cat_pres.drug == 'polyethyleneglycol','B'] = 1
cat_pres.loc[cat_pres.drug == 'neutraphos',['A','B']] = 1
cat_pres.loc[cat_pres.drug == 'simethicone','A'] = 1
cat_pres.loc[cat_pres.drug == 'potassiumphosphate','B'] = 1
cat_pres.loc[cat_pres.drug == 'phosphorus','B'] = 1
cat_pres.loc[cat_pres.drug == 'epoetinalfa','B'] = 1
cat_pres.loc[cat_pres.drug == 'sarnalotion', 'D'] = 1
cat_pres.loc[cat_pres.drug == 'readicatbariumsulfatesuspension',['A','V']] = 1 # Contrast agent
cat_pres.loc[cat_pres.drug == 'tiotropiumbromide','R'] = 1
cat_pres.loc[cat_pres.drug == 'sodiumpolystyrenesulfonate','V'] = 1
cat_pres.loc[cat_pres.drug == 'vasopressin','H'] = 1
cat_pres.loc[cat_pres.drug == 'fishoilmega','C'] = 1
cat_pres.loc[cat_pres.drug == 'ferricgluconate','B'] = 1
cat_pres.loc[cat_pres.drug == 'leucovorincalcium','V'] = 1
cat_pres.loc[cat_pres.drug == 'tuberculinprotein','V'] = 1
cat_pres.loc[cat_pres.drug == 'gastroviewdiatrizoatemegluminesodium',['A','V']] = 1 # Contrast agent
cat_pres.loc[cat_pres.drug == 'lithiumcarbonate','N'] = 1
cat_pres.loc[cat_pres.drug == 'gammagardliquid', 'J'] = 1
cat_pres.loc[cat_pres.drug == 'psylliumwafer', 'A'] = 1
cat_pres.loc[cat_pres.drug == 'caphosol', 'V'] = 1
cat_pres.loc[cat_pres.drug == 'vitaminbcomplex', 'B'] = 1
cat_pres.loc[cat_pres.drug == 'hydrocerin','D'] = 1
cat_pres.loc[cat_pres.drug == 'fosaprepitant','A'] = 1
cat_pres.loc[cat_pres.drug == 'psyllium', 'A'] = 1
cat_pres.loc[cat_pres.drug == 'levalbuterolneb','R'] = 1
cat_pres.loc[cat_pres.drug == 'aquaphorointment','D'] = 1
cat_pres.loc[cat_pres.drug == 'psylliumpowder','A'] = 1
cat_pres.loc[cat_pres.drug == 'protaminesulfate','V'] = 1
cat_pres.loc[cat_pres.drug == 'transplantselfmedicationprogram','V'] = 1
cat_pres.loc[cat_pres.drug == 'bismuthsubsalicylate','A'] = 1
cat_pres.loc[cat_pres.drug == 'xopenexneb','R'] = 1

In [45]:
cap_utils.rec_cat_filler(cat_pres,['gsn','drug','formulary_drug_cd','ndc'])

% of missing values before loop: 1.9826692207058676
gsn column split, starting loop...
Dataframe exploded, starting for loop #1

Loop #1, missing values % after filling with gsn column: 1.9789797142502779
Loop #1, missing values % after filling with drug column: 1.978748182011492
Loop #1, missing values % after filling with formulary_drug_cd column: 1.9779062465977257
Loop #1, missing values % after filling with ndc column: 1.9700604609606913

Loop #1, missing values % after re-imploding df: 1.9563567427505484
Dataframe exploded, starting for loop #2

Loop #2, missing values % after filling with gsn column: 1.9684607836745356
Loop #2, missing values % after filling with drug column: 1.9684502594818636
Loop #2, missing values % after filling with formulary_drug_cd column: 1.9684502594818636
Loop #2, missing values % after filling with ndc column: 1.9684502594818636

Loop #2, missing values % after re-imploding df: 1.9549479936801513
Dataframe exploded, starting for loop #3

Loop #3, mis

In [46]:
catnas = cat_pres[cat_pres.A < 0]
catnas[['ndc','drug_type','drug','formulary_drug_cd','gsn']].isna().sum()/len(catnas)*100

ndc                   7.422661
drug_type            94.407499
drug                 96.615513
formulary_drug_cd    96.615513
gsn                  94.921110
dtype: float64

In [47]:
cap_utils.display_sbs(['ndc value counts']*2,\
            cat_pres.ndc[cat_pres.A<0].value_counts(normalize = True).iloc[0:25]*100,\
            cat_pres.ndc[cat_pres.A<0].value_counts(normalize = True).iloc[25:50]*100)
missing_ndcs = cat_pres.ndc[cat_pres.A<0].value_counts(normalize = True).head(30)
print(f'We could fill {round(missing_ndcs.sum()*100,2)}% of the remaining missing ndcs manually by filling the first 30 ndcs')

ndc value counts 
 
 
   
 ndc 
 
 
 
 
 100030116 
 7.881357 
 
 
 105850865 
 6.977598 
 
 
 100030115 
 6.080837 
 
 
 105720810 
 5.019358 
 
 
 107700128 
 4.967173 
 
 
 353560215 
 2.316245 
 
 
 109390331 
 1.737257 
 
 
 109390803 
 1.670787 
 
 
 123330025 
 1.592656 
 
 
 109390021 
 1.275757 
 
 
 328490500 
 1.176635 
 
 
 100390049 
 1.126783 
 
 
 103100000 
 1.081303 
 
 
 109390170 
 0.977517 
 
 
 118222280 
 0.817464 
 
 
 328490706 
 0.744872 
 
 
 118225764 
 0.740207 
 
 
 103270035 
 0.711637 
 
 
 109160046 
 0.699976 
 
 
 101350203 
 0.658286 
 
 
 109390232 
 0.652164 
 
 
 614420501 
 0.635546 
 
 
 109390026 
 0.618346 
 
 
 328490707 
 0.612515 
 
 
 122800008 
 0.593857 
 
 
 

 
 ndc value counts 
 
 
   
 ndc 
 
 
 
 
 109390197 
 0.547503 
 
 
 461220143 
 0.499399 
 
 
 107540669 
 0.499399 
 
 
 101350120 
 0.474327 
 
 
 109390378 
 0.447215 
 
 
 109390789 
 0.442259 
 
 
 118220148 
 0.442259 
 
 
 116940846 
 0.426516 
 
 
 109390556 
 0.413105 
 
 
 101190021 
 0.399111 
 
 
 107120478 
 0.397654 
 
 
 106310098 
 0.396488 
 
 
 118450060 
 0.392115 
 
 
 108760691 
 0.391531 
 
 
 101350683 
 0.387158 
 
 
 118224607 
 0.386575 
 
 
 122800077 
 0.382202 
 
 
 109390008 
 0.377246 
 
 
 370000143 
 0.371124 
 
 
 328490708 
 0.366168 
 
 
 110262618 
 0.364127 
 
 
 543383 
 0.346344 
 
 
 499990990 
 0.334099 
 
 
 102670737 
 0.331475 
 
 
 103700511 
 0.325353

We could fill 53.83% of the remaining missing ndcs manually by filling the first 30 ndcs


In [49]:
prescriptions

join_on_index  subject_id   hadm_id            starttime  \
ndc                                                                     
904198861                0    10000032  22595853  2180-05-07 01:00:00   
0                        1    10000032  22595853  2180-05-07 00:00:00   
51079007320              2    10000032  22595853  2180-05-08 08:00:00   
6022761                  3    10000032  22595853  2180-05-07 01:00:00   
63323026201              4    10000032  22595853  2180-05-07 00:00:00   
...                    ...         ...       ...                  ...   
12280027016       18952980    19999828  32917002  2149-01-08 09:41:00   
11845110101       18952981    19999828  32917002  2149-01-08 09:41:00   
10544007120       18952982    19999828  32917002  2149-01-08 09:41:00   
10544093814       18952983    19999828  32917002  2149-01-08 09:41:00   
10544034130       18952984    19999828  32917002  2149-01-08 09:41:00   

            drug_type                         drug formulary_drug_cd     gsn  \
ndc                                                                            
904198861        MAIN                Acetaminophen           APAP500  004490   
0                MAIN  Sodium Chloride 0.9%  Flush         NACLFLUSH     NaN   
51079007320      MAIN                   Furosemide            FURO40  008209   
6022761          MAIN                  Raltegravir           RALT400  063231   
63323026201      MAIN                      Heparin            HEPA5I  006549   
...               ...                          ...               ...     ...   
12280027016       NaN                  fluticasone               NaN   18368   
11845110101       NaN                    glipizide               NaN    1776   
10544007120       NaN                    ibuprofen               NaN    8350   
10544093814       NaN                metronidazole               NaN    9592   
10544034130       NaN                   omeprazole               NaN   33530   

             route  stay_id_ind  
ndc                              
904198861    PO/NG            0  
0               IV            0  
51079007320  PO/NG            0  
6022761         PO            0  
63323026201     SC            0  
...            ...          ...  
12280027016    NaN            1  
11845110101    NaN            1  
10544007120    NaN            1  
10544093814    NaN            1  
10544034130    NaN            1  

[18952985 rows x 10 columns]

In [48]:
cat_pres

ndc  join_on_index drug_type                 drug  \
0         <NA>              1      MAIN  sodiumchlorideflush   
1         <NA>             22      MAIN  sodiumchlorideflush   
2         <NA>             39      BASE       sodiumchloride   
3         <NA>             50      MAIN  sodiumchlorideflush   
4         <NA>             51      BASE       sodiumchloride   
...        ...            ...       ...                  ...   
18952980  <NA>       16216912      MAIN     magnesiumsulfate   
18952981  <NA>       16216929      MAIN      atropinesulfate   
18952982  <NA>       16216941      MAIN        phenylephrine   
18952983  <NA>       16216977      MAIN             propofol   
18952984  <NA>       16216982      MAIN       norepinephrine   

         formulary_drug_cd  gsn  A  B  C  D  G  H  J  L  M  N  P  R  S  V  
0                NACLFLUSH  NaN  1  1  0  0  0  0  0  0  0  0  0  0  0  0  
1                NACLFLUSH  NaN  1  1  0  0  0  0  0  0  0  0  0  0  0  0  
2                  NS100CG  NaN  1  1  0  0  0  0  0  0  0  0  0  0  0  0  
3                NACLFLUSH  NaN  1  1  0  0  0  0  0  0  0  0  0  0  0  0  
4                  NS100CG  NaN  1  1  0  0  0  0  0  0  0  0  0  0  0  0  
...                    ...  ... .. .. .. .. .. .. .. .. .. .. .. .. .. ..  
18952980            MAG2PM  NaN  1  1  0  1  0  0  0  0  0  0  0  0  0  1  
18952981            ATRO1I  NaN  1  0  0  0  0  0  0  0  0  0  0  0  1  0  
18952982           PHEN10I  NaN  0  0  1  0  0  0  0  0  0  0  0  1  1  0  
18952983         PROP100IG  NaN  0  0  0  0  0  0  0  0  0  1  0  0  0  0  
18952984         NORE8/250  NaN  0  0  1  0  0  0  0  0  0  0  0  0  0  0  

[18952985 rows x 20 columns]

The problem is that the missing values are more distributed between many ndcs, so filling manually 30 of them would just tackle 53.83% of the missing categories by ndc. Moreover, browsing ndc is not so easy, as they are specific per supplier and can be harder to find. Even more if they could not be found on the FDA website. We will leave these values unfilled for now. We can save the merged, filled prescriptions df to be used later.

In [52]:
cat_pres

ndc  join_on_index drug_type                 drug  \
0         <NA>              1      MAIN  sodiumchlorideflush   
1         <NA>             22      MAIN  sodiumchlorideflush   
2         <NA>             39      BASE       sodiumchloride   
3         <NA>             50      MAIN  sodiumchlorideflush   
4         <NA>             51      BASE       sodiumchloride   
...        ...            ...       ...                  ...   
18952980  <NA>       16216912      MAIN     magnesiumsulfate   
18952981  <NA>       16216929      MAIN      atropinesulfate   
18952982  <NA>       16216941      MAIN        phenylephrine   
18952983  <NA>       16216977      MAIN             propofol   
18952984  <NA>       16216982      MAIN       norepinephrine   

         formulary_drug_cd  gsn  A  B  C  D  G  H  J  L  M  N  P  R  S  V  
0                NACLFLUSH  NaN  1  1  0  0  0  0  0  0  0  0  0  0  0  0  
1                NACLFLUSH  NaN  1  1  0  0  0  0  0  0  0  0  0  0  0  0  
2                  NS100CG  NaN  1  1  0  0  0  0  0  0  0  0  0  0  0  0  
3                NACLFLUSH  NaN  1  1  0  0  0  0  0  0  0  0  0  0  0  0  
4                  NS100CG  NaN  1  1  0  0  0  0  0  0  0  0  0  0  0  0  
...                    ...  ... .. .. .. .. .. .. .. .. .. .. .. .. .. ..  
18952980            MAG2PM  NaN  1  1  0  1  0  0  0  0  0  0  0  0  0  1  
18952981            ATRO1I  NaN  1  0  0  0  0  0  0  0  0  0  0  0  1  0  
18952982           PHEN10I  NaN  0  0  1  0  0  0  0  0  0  0  0  1  1  0  
18952983         PROP100IG  NaN  0  0  0  0  0  0  0  0  0  1  0  0  0  0  
18952984         NORE8/250  NaN  0  0  1  0  0  0  0  0  0  0  0  0  0  0  

[18952985 rows x 20 columns]

In [51]:
prescriptions

join_on_index  subject_id   hadm_id            starttime  \
ndc                                                                     
904198861                0    10000032  22595853  2180-05-07 01:00:00   
0                        1    10000032  22595853  2180-05-07 00:00:00   
51079007320              2    10000032  22595853  2180-05-08 08:00:00   
6022761                  3    10000032  22595853  2180-05-07 01:00:00   
63323026201              4    10000032  22595853  2180-05-07 00:00:00   
...                    ...         ...       ...                  ...   
12280027016       18952980    19999828  32917002  2149-01-08 09:41:00   
11845110101       18952981    19999828  32917002  2149-01-08 09:41:00   
10544007120       18952982    19999828  32917002  2149-01-08 09:41:00   
10544093814       18952983    19999828  32917002  2149-01-08 09:41:00   
10544034130       18952984    19999828  32917002  2149-01-08 09:41:00   

            drug_type                         drug formulary_drug_cd     gsn  \
ndc                                                                            
904198861        MAIN                Acetaminophen           APAP500  004490   
0                MAIN  Sodium Chloride 0.9%  Flush         NACLFLUSH     NaN   
51079007320      MAIN                   Furosemide            FURO40  008209   
6022761          MAIN                  Raltegravir           RALT400  063231   
63323026201      MAIN                      Heparin            HEPA5I  006549   
...               ...                          ...               ...     ...   
12280027016       NaN                  fluticasone               NaN   18368   
11845110101       NaN                    glipizide               NaN    1776   
10544007120       NaN                    ibuprofen               NaN    8350   
10544093814       NaN                metronidazole               NaN    9592   
10544034130       NaN                   omeprazole               NaN   33530   

             route  stay_id_ind  
ndc                              
904198861    PO/NG            0  
0               IV            0  
51079007320  PO/NG            0  
6022761         PO            0  
63323026201     SC            0  
...            ...          ...  
12280027016    NaN            1  
11845110101    NaN            1  
10544007120    NaN            1  
10544093814    NaN            1  
10544034130    NaN            1  

[18952985 rows x 10 columns]

In [62]:
cat.append('join_on_index')
cat_pres = cat_pres[cat].set_index('join_on_index',drop = True)
prescriptions = prescriptions[['subject_id','hadm_id','starttime','stay_id_ind','join_on_index']].set_index('join_on_index', drop = True)
prescriptions_final = prescriptions.join(cat_pres)

In [65]:
prescriptions_final.to_csv('prescriptions_final.csv')

In [ ]:
non_null_cats = cat_pres.A>=0
cat_pres = cat_pres.loc[non_null_cats,cat]
prescriptions = prescriptions[['subject_id','hadm_id','starttime']]
prescriptions = prescriptions.join(cat_pres)
prescriptions.reset_index(drop = True, inplace = True)

In [ ]:
cat_pres.A.value_counts(normalize = True)*100

In [ ]:
unclass_drugs = cat_pres.drug[cat_pres.A<0].drop_duplicates()
udrugs_gsnvc = cat_pres.gsn[cat_pres.A<0].value_counts()
gsn_udrugs = udrugs_gsnvc[udrugs_gsnvc>1].index.values

In [ ]:
cat.append('gsn')
cat_gsn = cat_pres.loc[cat_pres.gsn.isin(gsn_udrugs),cat].groupby('gsn').max()
cat_gsn = cat_gsn[cat_gsn.A>-1]
cat_gsn

In [ ]:
cat = cat[:-1]
cat_pres[cat] = cat_pres[cat].where(cat_pres[cat]>=0)

In [ ]:
cat_pres.set_index(ind,append = True)

In [ ]:
cat_pres.set_index('gsn', inplace = True)
cat_pres.update(cat_gsn)
cat_pres.set_index(ind)

In [ ]:
cat_pres.head()

In [ ]:
sum(cat_pres.drug.str.contains('polyvalent'))

In [ ]:
cat_pres.gsn[misspelt_drugs_bool].value_counts().describe()

In [ ]:
(5303415+7791711)/len(cat_pres)

In [ ]:
cat

In [ ]:
cat = list('ABCDGHJLMNPRSV')
cat.append('drug')
cat_drug = cat_pres[cat].fillna(0.0001).groupby('formulary_drug_cd',dropna = False).sum()
cat_drug

In [ ]:
atr4_pres.atc4_name.nunique()

The missing values labelled as 0 are now empty strings due to the conversion. There may be other values that are not 0, maybe even some could be in other ndc formats (hyphenated or with less digits). We can find out how many rows have which number of digits.

So all fields that do not have 9-digit numbers have no digits. We can consider all of these as missing values.

More than 12% of ndc fields are missing. Some of the drugs listed here may not have an ndc (Sodium Chloride flush on 2nd row for instance), but others may be mistakes. NDC is specific to product, not to drug (same drug from different companies will have a different code) so filling from the table would not be very accurate. Given that what we need the NDC just to classify our drugs, we will leave these fields blank and will fill the categories instead after NDC-matching.


In [ ]:
prescriptions.ndc[~missing_ndc].drop_duplicates().to_csv('Unique_NDCs.csv')

Let us see if we can fill these missing values by using the other fields in the table. pharmacy_id, gsn, drug and formulary_drug_cd could be used for this.

In [ ]:
prescriptions.formulary_drug_cd.info()

In [ ]:
fdrugna = prescriptions[prescriptions.formulary_drug_cd.isna()]

In [ ]:
def df_filler(x):
    

prescriptions.formulary_drug_cd[prescriptions.drug.isin(fdrugna.drug)].isna()

In [ ]:
prescriptions.gsn.value_counts()

In [ ]:
prescriptions.pharmacy_id.value_counts()

In [ ]:
prescriptions.formulary_drug_cd.value_counts()

In [ ]:
prescriptions.formulary_drug_cd.info()

In [ ]:
prescriptions.drug = prescriptions.drug.str.lower()
prescriptions.drug.value_counts().describe()

In [ ]:
ndc_bool = prescriptions.ndc!=u'' & prescriptions 
print(sum(ndc_bool))
print(sum(~ndc_bool))

In [ ]:
elswhr_bool = prescriptions.loc[~ndc_bool,'drug'].isin(prescriptions.loc[ndc_bool,'drug'])
print(round(sum(elswhr_bool)/sum(~ndc_bool)*100,2))

In [ ]:
prescriptions.loc[elswhr_bool,'ndc'] = prescriptions.loc[ndc_bool*prescriptions.,'ndc']

In [ ]:
prescriptions.ndc.nunique()

In [ ]:
import numpy as np

In [ ]:
medication = pd.read_json(r'C:\Users\Nacho\Downloads\Medication.ndjson', lines = True)

In [ ]:
medication.shape

In [ ]:
prescriptions.ndc.head()

In [ ]:
ndc_desc = pd.read_csv("G:\My Drive\Brainstation\Capstone project\product.csv",usecols = ['PRODUCTNDC','NONPROPRIETARYNAME','PHARM_CLASSES'])

In [ ]:
ndc_desc = ndc_desc.loc[~ndc_desc.PHARM_CLASSES.isna()].reset_index(drop = True)

In [ ]:
ndc_desc.head()

In [ ]:
def sht2nineNDC(x):
    x = x.split('-')
    x = x[0].zfill(5) + x[1].zfill(4)
    
    return x
    

ndc_desc.PRODUCTNDC = ndc_desc.PRODUCTNDC.apply(sht2nineNDC)

In [ ]:
ndc_desc.PRODUCTNDC.head()

In [ ]:
classpername = ndc_desc[['NONPROPRIETARYNAME','PHARM_CLASSES']].groupby('NONPROPRIETARYNAME').nunique()

In [ ]:
classperndc = ndc_desc[['PRODUCTNDC','PHARM_CLASSES']].groupby('PRODUCTNDC').nunique()

In [ ]:
classpername.value_counts()

In [ ]:
classperndc.value_counts()

In [ ]:
np.mean(prescriptions.ndc.isna())*100

In [ ]:
prescriptions.loc[prescriptions.ndc.isna()].head(10)

In [ ]:
ndc_desc.rename(columns = {'PRODUCTNDC' : 'ndc'},inplace = True)

In [ ]:
presfinal = pd.merge(prescriptions,ndc_desc, how = 'left',on = 'ndc')

In [ ]:
np.mean(presfinal.PHARM_CLASSES.isna())

In [ ]:
np.mean(presfinal.NONPROPRIETARYNAME.isna())

In [ ]:
presfinal[['ndc','drug','NONPROPRIETARYNAME','PHARM_CLASSES']]

In [ ]:
presfinal.hadm_id.nunique()

In [ ]:
len(presfinal.PHARM_CLASSES.isna())-sum(presfinal.PHARM_CLASSES.isna())

In [ ]:
classperhadm = presfinal[['PHARM_CLASSES','hadm_id']].groupby('hadm_id').nunique()
classperhadm.describe()

In [ ]:
classperhadm.loc[classperhadm.PHARM_CLASSES == 0].count()

In [ ]:
prescriptions.ndc.apply(lambda x: str(x)[:-2].zfill(9))

In [ ]:
print(ndc_desc.NONPROPRIETARYNAME.nunique())
print(ndc_desc_clean.NONPROPRIETARYNAME.nunique())

In [ ]:
ndc_desc.shape

In [ ]:
ndc_desc.PRODUCTNDC.apply(lambda x: int(x.replace('-','')))

In [ ]:
ndc_desc.PHARM_CLASSES.value_counts

In [ ]:
poe = pd.read_csv('MIMIC IV ED/poe.csv')

In [ ]:
poe.order_status.value_counts(normalize = True)

In [ ]:
poe.loc[poe.order_type == 'Lab'].order_subtype.value_counts()

In [ ]:
poe.order_type.value_counts()

In [ ]:
admissions = pd.read_csv('MIMIC IV ED/admissions.csv')

In [ ]:
admissions.race.value_counts(normalize = True)

In [ ]:
sum(admissions.admission_location.isna())

In [ ]:
admissions.admission_location.value_counts(normalize = True)

In [ ]:
transfers = pd.read_csv('MIMIC IV ED/transfers.csv')

In [ ]:
sum(transfers.eventtype == 'discharge')

In [ ]:
transfers.careunit.value_counts(normalize = True)

In [ ]:
transfers.careunit.unique()

In [ ]:
transfers[['intime','outtime']] = pd.to_datetime(transfers[['intime', 'outtime']])

In [ ]:
transfers.info(show_counts = True)

In [ ]:
bool_outs

In [ ]:
transfers.describe()

In [ ]:
bool_discharge = transfers.eventtype=='discharge'
bool_carenan = transfers.careunit.isna()
bool_outtime = transfers.outtime.isna()

In [ ]:
print(all(bool_discharge == bool_carenan))
print(all(bool_discharge == bool_outtime))


In [ ]:
transfers.head()

In [ ]:
services = pd.read_csv('MIMIC IV ED/services.csv')

In [ ]:
services.shape

In [ ]:
services.head()

In [ ]:
services.prev_service.value_counts()

In [ ]:
transfers.eventtype.value_counts()

In [ ]:
transfers.

In [ ]:
emar = pd.read_csv('MIMIC IV/emar.csv')

In [ ]:
emar.medication.nunique()

In [ ]:
emar.head()

In [ ]:
transfers.careunit.value_counts()

In [ ]:
print('Number of patients:',transfers['subject_id'].nunique())
print('Number of transfers (and rows number):',transfers['transfer_id'].nunique())
print('Number of stays:',transfers['hadm_id'].nunique())
print('Number of departments:',transfers['careunit'].nunique())

In [ ]:
sum(transfers['careunit'].isna())

In [ ]:
print('Number of departments:',transfers['careunit'].unique())

In [ ]:
admissions = pd.read_csv('MIMIC IV ED/admissions.csv')

In [ ]:
pd.to_datetime(admissions.admittime,infer_datetime_format = True).describe(datetime_is_numeric = True)

In [ ]:
admissions.discharge_location.unique()

In [ ]:
admissions.head()

In [ ]:
len(admissions)

In [ ]:
print('Number of patients:',admissions['subject_id'].nunique())
print('Number of transfers (and length of rows):',admissions['transfer_id'].nunique())
print('Number of stays:',transfers['hadm_id'].nunique())
print('Number of departments:',transfers['careunit'].nunique())

In [ ]:
edstays['disposition'].value_counts()

In [ ]:
diagnosis.icd_version.value_counts()

In [ ]:
diagnosis['icd_title'].value_counts().head()

In [ ]:
print('\033[1m' + 'diagnosis')
display(diagnosis.head())
display(diagnosis.info())
print('\033[1m' + 'edstays')
display(edstays.head())
display(diagnosis.info())
print('\033[1m' + 'medrecon')
display(medrecon.head())
display(medrecon.info())
print('\033[1m' + 'pyxis')
display(pyxis.head())
display(pyxis.info())
print('\033[1m' + 'triage')
display(triage.head())
display(triage.info())
print('\033[1m' + 'vitalsign')
display(vitalsign.head())
display(vitalsign.info())